File: app/src/main/java/im/tox/antox/tox/ToxDoService.java
package im.tox.antox.tox;

import android.app.Notification;
import android.app.PendingIntent;
import android.app.Service;
import android.content.Context;
import android.content.Intent;
import android.os.IBinder;
import android.os.PowerManager;
import android.support.v4.app.NotificationCompat;
import android.util.Log;

import im.tox.antox.R;
import im.tox.antox.activities.MainActivity;
import im.tox.jtoxcore.ToxException;

public class ToxDoService extends Service {

    private static final String TAG = "im.tox.antox.tox.ToxDoService";

    private static ToxSingleton toxSingleton = ToxSingleton.getInstance();
    private PowerManager.WakeLock serviceWL;
    private Context ctx;

    public ToxDoService() {
        super();
    }

    @Override
    public void onCreate() {
        this.ctx = getApplicationContext();
        Log.d(TAG, "onCreate");
        if (!toxSingleton.isInited) {
            toxSingleton.initTox(this.ctx);
            Log.d(TAG, "Initting ToxSingleton");
        }

        PowerManager pm = (PowerManager) getSystemService(Context.POWER_SERVICE);
        this.serviceWL =
                pm.newWakeLock(PowerManager.PARTIAL_WAKE_LOCK, "Antox Service Wakelock");

        Intent resultIntent = new Intent(this, MainActivity.class);
        PendingIntent pendingIntent =
                PendingIntent.getActivity(ctx, 0, resultIntent, PendingIntent.FLAG_UPDATE_CURRENT);

        NotificationCompat.Builder mBuilder =
                new NotificationCompat.Builder(this)
                        .setSmallIcon(R.drawable.ic_actionbar)
                        .setContentTitle("Antox service is running")
                        .setDefaults(Notification.DEFAULT_ALL)
                        .setContentIntent(pendingIntent);

        Thread doThread = new Thread() {
            @Override
            public void run() {
                /* Praise the sun */
                while (true) {
                    try {
                        Thread.sleep(toxSingleton.jTox.doToxInterval());
                        toxSingleton.jTox.doTox();
                    } catch (ToxException e) {
                        Log.e(TAG, e.getError().toString());
                        e.printStackTrace();
                    } catch (InterruptedException e) {
                        Log.e(TAG, e.getMessage());
                        e.printStackTrace();
                    }
                }
            }
        };
        serviceWL.acquire();
        startForeground(-1, mBuilder.build());
        doThread.start();
    }

    @Override
    public IBinder onBind(Intent intent) {
        return null;
    }

    @Override
    public int onStartCommand(Intent intent, int flags, final int id) {
        return START_NOT_STICKY;
    }

    @Override
    public void onDestroy() {
        Log.d(TAG, "onDestroy() called");
        stopForeground(true);
        toxSingleton.isInited = false;
        try {
            toxSingleton.jTox.killTox();
        } catch (ToxException e) {
            e.printStackTrace();
        }
        toxSingleton = null;
        serviceWL.release();
        super.onDestroy();
    }
}

File: app/src/main/scala/im/tox/antox/tox/ToxDoService.scala
/*
 * Copyright (c) 2014 Mark Winter (Astonex)
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
 * MA  02110-1301, USA.
 */


package im.tox.antox.tox

import android.app.Service
import android.content.{Context, Intent}
import android.net.ConnectivityManager
import android.os.IBinder
import android.preference.PreferenceManager
import android.util.Log

class ToxDoService extends Service() {

  private var serviceThread: Thread = _

  private var keepRunning: Boolean = true

  override def onCreate() {
    if (!ToxSingleton.isInited) {
      ToxSingleton.initTox(getApplicationContext)
      Log.d("ToxDoService", "Initting ToxSingleton")
    }
    keepRunning = true
    val start = new Runnable() {

      override def run() {
        while (keepRunning) {
          val preferences = PreferenceManager.getDefaultSharedPreferences(getApplicationContext)
          val connManager = getSystemService(Context.CONNECTIVITY_SERVICE).asInstanceOf[ConnectivityManager]
          val wifiOnly = preferences.getBoolean("wifi_only", true)
          val mWifi = connManager.getNetworkInfo(ConnectivityManager.TYPE_WIFI)
          if (wifiOnly && !mWifi.isConnected) {
            try {
              Thread.sleep(10000)
            } catch {
              case e: Exception =>
            }
          } else {
            try {
              Thread.sleep(ToxSingleton.jTox.doToxInterval())
              ToxSingleton.jTox.doTox()
            } catch {
              case e: Exception =>
            }
          }
        }
      }
    }
    serviceThread = new Thread(start)
    serviceThread.start()
  }

  override def onBind(intent: Intent): IBinder = null

  override def onStartCommand(intent: Intent, flags: Int, id: Int): Int = Service.START_STICKY

  override def onDestroy() {
    super.onDestroy()
    keepRunning = false
    serviceThread.interrupt()
    ToxSingleton.isInited = false
    Log.d("ToxDoService", "onDestroy() called")
  }
}
